features/TESTS/filesystem/util_block_device/main.cpp

```diff
@@ -0,0 +1,171 @@
+#include "mbed.h"
+#include "greentea-client/test_env.h"
+#include "unity.h"
+#include "utest.h"
+
+#include "HeapBlockDevice.h"
+#include "SlicingBlockDevice.h"
+#include "ChainingBlockDevice.h"
+#include <stdlib.h>
+
+using namespace utest::v1;
+
+#define BLOCK_SIZE 512
+uint8_t write_block[BLOCK_SIZE];
+uint8_t read_block[BLOCK_SIZE];
+
+
+void test_slicing() {
+    HeapBlockDevice bd(16*BLOCK_SIZE, BLOCK_SIZE);
+
+    // Test with first slice of block device
+    SlicingBlockDevice slice1(&bd, 0, 8*BLOCK_SIZE);
+
+    int err = slice1.init();
+    TEST_ASSERT_EQUAL(0, err);
+
+    TEST_ASSERT_EQUAL(BLOCK_SIZE, slice1.write_size());
+    TEST_ASSERT_EQUAL(8*BLOCK_SIZE, slice1.size());
+
+    // Fill with random sequence
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        write_block[i] = 0xff & rand();
+    }
+
+    // Write, sync, and read the block
+    err = slice1.write(write_block, 0, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    err = slice1.read(read_block, 0, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    // Check that the data was unmodified
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i]);
+    }
+
+    // Check with original block device
+    err = bd.read(read_block, 0, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    // Check that the data was unmodified
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i]);
+    }
+
+    err = slice1.deinit();
+    TEST_ASSERT_EQUAL(0, err);
+
+
+    // Test with second slice of block device
+    SlicingBlockDevice slice2(&bd, -8*BLOCK_SIZE);
+
+    err = slice2.init();
+    TEST_ASSERT_EQUAL(0, err);
+
+    TEST_ASSERT_EQUAL(BLOCK_SIZE, slice2.write_size());
+    TEST_ASSERT_EQUAL(8*BLOCK_SIZE, slice2.size());
+
+    // Fill with random sequence
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        write_block[i] = 0xff & rand();
+    }
+
+    // Write, sync, and read the block
+    err = slice2.write(write_block, 0, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    err = slice2.read(read_block, 0, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    // Check that the data was unmodified
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i]);
+    }
+
+    // Check with original block device
+    err = bd.read(read_block, 8*BLOCK_SIZE, BLOCK_SIZE);
+    TEST_ASSERT_EQUAL(0, err);
+
+    // Check that the data was unmodified
+    srand(1);
+    for (int i = 0; i < BLOCK_SIZE; i++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i]);
+    }
+
+    err = slice2.deinit();
+    TEST_ASSERT_EQUAL(0, err);
+}
+
+void test_chaining() {
+    HeapBlockDevice bd1(8*BLOCK_SIZE, BLOCK_SIZE);
```

Comment:
The constant 8 is all over the file, shall it be ``const uint8_t blocks_number = 8;`` for instance? I would also use a regular variable rather than a macro for BLOCK_SIZE if possible